pkg/provider/cloud/kubevirt/data_volume.go

```diff
@@ -23,49 +23,121 @@ import (
 	utilpointer "k8s.io/utils/pointer"
 	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+const (
+	preAllocatedDVAnnotation  = "cdi.kubevirt.io/storage.deleteAfterCompletion"
+	preAllocatedDVSize        = "11Gi"
+	osAnnotationForCustomDisk = "cdi.kubevirt.io/os-type"
+	dataVolumeClusterRole     = "datavolume-cloner"
+	dataVolumeRoleBinding     = "allow-datavolume-cloning"
+)
+
 func dataVolumeCreator(datavolume *cdiv1beta1.DataVolume) reconciling.NamedCDIv1beta1DataVolumeCreatorGetter {
 	return func() (name string, create reconciling.CDIv1beta1DataVolumeCreator) {
 		return datavolume.Name, func(dv *cdiv1beta1.DataVolume) (*cdiv1beta1.DataVolume, error) {
+			dv.Annotations = datavolume.Annotations
 			dv.Spec = datavolume.Spec
 			return dv, nil
 		}
 	}
 }
 
-func reconcilePreAllocatedDataVolumes(ctx context.Context, cluster *kubermaticv1.Cluster, client ctrlruntimeclient.Client) error {
+func reconcileDataVolume(ctx context.Context, client ctrlruntimeclient.Client, dataVolume *cdiv1beta1.DataVolume, namespace string) error {
+	dvCreator := []reconciling.NamedCDIv1beta1DataVolumeCreatorGetter{
+		dataVolumeCreator(dataVolume),
+	}
+	return reconciling.ReconcileCDIv1beta1DataVolumes(ctx, dvCreator, namespace, client)
+}
+
+// reconcileLocalDataVolumes reconciles the local-disks from cluster.
+func reconcileLocalDataVolumes(ctx context.Context, cluster *kubermaticv1.Cluster, client ctrlruntimeclient.Client) error {
 	for _, d := range cluster.Spec.Cloud.Kubevirt.PreAllocatedDataVolumes {
-		dv, err := createPreAllocatedDataVolume(d, cluster.Status.NamespaceName)
+		dv, err := newDataVolume(d, cluster.Status.NamespaceName)
 		if err != nil {
 			return err
 		}
-		dvCreator := []reconciling.NamedCDIv1beta1DataVolumeCreatorGetter{
-			dataVolumeCreator(dv),
+		if err := reconcileDataVolume(ctx, client, dv, cluster.Status.NamespaceName); err != nil {
+			return fmt.Errorf("failed to reconcile DataVolume: %w", err)
+		}
+	}
+	return nil
+}
+
+// reconcilePreAllocatedDataVolume reconciles the preAllocated DataVolumes from standard VM images if cloning is enabled.
+func reconcilePreAllocatedDataVolume(ctx context.Context, dc *kubermaticv1.DatacenterSpecKubevirt, client ctrlruntimeclient.Client) error {
+	if dc.Images == nil || dc.Images.HTTP == nil || dc.Images.HTTP.ImageCloning == nil || !dc.Images.HTTP.ImageCloning.Enable {
+		return nil
+	}
+	existingDiskList := cdiv1beta1.DataVolumeList{}
+	listOption := ctrlruntimeclient.ListOptions{
+		Namespace: KubevirtDiskCacheNamespace,
+	}
+	if err := client.List(ctx, &existingDiskList, &listOption); ctrlruntimeclient.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// pdvToBeRemoved contains info about pdv that are going to be removed.
+	pdvToBeRemoved := make(map[string]cdiv1beta1.DataVolume)
+	for _, dv := range existingDiskList.Items {
+		// only consider preAllocated DV and ignore custom-disks.
+		if _, exist := dv.Annotations[osAnnotationForCustomDisk]; !exist {
```

Comment:
How do we ensure that `osAnnotationForCustomDisk` will be present for custom disks ?
Would it be safer to have an annotation for what we own (PreAllocatedDataVolumes) and reconcile what we own ?
If the user does not set this annotation (very likeyly, as it's based on documentation only if I understood correctly), it would be removed.
I find safer to annotate our own DV and reconcile what has our annotation (and we can ensure via the reconcile here that it has the correct annotation).

Reply:
+1.

I will add another annotation to our own reconciled PDVs.